rpc/database.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Athanor Labs (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package rpc
+
+import (
+	"math/big"
+	"net/http"
+
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	"github.com/athanorlabs/atomic-swap/db"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+// RecoveryDB contains methods for retrieving swap recovery info from the database.
+type RecoveryDB interface {
+	GetContractSwapInfo(id types.Hash) (*db.EthereumSwapInfo, error)
+	GetSwapPrivateKey(id types.Hash) (*mcrypto.PrivateSpendKey, error)
+	GetCounterpartySwapPrivateKey(id types.Hash) (*mcrypto.PrivateSpendKey, error)
+}
+
+// DatabaseService ...
+type DatabaseService struct {
+	rdb RecoveryDB
+}
+
+// NewDatabaseService returns a new DatabaseService.
+func NewDatabaseService(rdb RecoveryDB) *DatabaseService {
+	return &DatabaseService{
+		rdb: rdb,
+	}
+}
+
+// GetContractSwapInfoRequest ...
+type GetContractSwapInfoRequest struct {
+	ID types.Hash `json:"id" validate:"required"`
+}
+
+// GetContractSwapInfoResponse ...
+type GetContractSwapInfoResponse struct {
+	StartNumber     *big.Int                   `json:"startNumber" validate:"required"`
+	SwapID          types.Hash                 `json:"swapID" validate:"required"`
+	Swap            *contracts.SwapCreatorSwap `json:"swap" validate:"required"`
```

Comment:
Why are we passing the `SwapID` if we are also passing the object that the `SwapID` can be quickly derived from?

Reply:
the user would need to manually write code to hash the struct to check their on-chain status which is bad

Reply:
Apologies. I was confused thinking this was a request when I wrote that. I'll resolve this.